### ISSUE TITLE
Highlighting fails with MoreLikeThis query and unified highlighter

### ISSUE THREAD

**Issue by sciphilo** [elastic/elasticsearch#26357]
Elasticsearch v5.5.0 fails to highlight with an MLT query when the unified highlighter is used.

**Elasticsearch version** (`bin/elasticsearch --version`):
5.5.0
**Plugins installed**:  default
**JVM version** (`java -version`):
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):

Darwin 15.6.0 Darwin Kernel Version 15.6.0: Sun Jun  4 21:43:07 PDT 2017; root:xnu-3248.70.3~1/RELEASE_X86_64 x86_64

MacOS 10.11.6

**Description of the problem**:

When running a More Like This query the unified highlighter doesn't work.  
The plain highlighter works, but not the unified one- there is simply no highlighting with unified.

**Steps to reproduce**:

 1. create index
    curl -XPUT 'http://localhost:9200/test'

 2.  Load data
    for id in `seq 1 3`; do  curl -XPUT http://localhost:9200/test/x/${id} --upload-file ${id}.json; done
 
3. Control

```
curl -XPOST 'localhost:9200/test/x/_search?pretty' -d '{
  "from" : 0,
  "size" : 1,
  "query" : {
               "bool" : {
                        "must" : [
                          {
                            "more_like_this" : {
                              "fields" : [
                                "data"
                              ],
                              "like" : [
                                {
                                  "_index" : "test",
                                  "_type" : "x",
                                  "_id" : "1"
                                }
                              ],
                              "max_query_terms" : 50,
                              "min_term_freq" : 1,
                              "min_doc_freq" : 1
                             
                            }
                          }
                        ],
                        "disable_coord" : false,
                        "adjust_pure_negative" : true,
                        "boost" : 1.0
                      }
              },
  "highlight" : {
    "pre_tags" : [
      "<highlight>"
    ],
    "post_tags" : [
      "</highlight>"
    ],
    "number_of_fragments" : 0,
    "type" : "plain",
    "fields" : {
 
      "data" : { }
   
    }
  }

}'
```

Note highlighting:

```
"hits" : [
      {
        "_index" : "test",
        "_type" : "x",
        "_id" : "2",
        "_score" : 1.1433042,
        "_source" : {
          "data" : "some data to highlight that is different"
        },
        "highlight" : {
          "data" : [
            "<highlight>some</highlight> <highlight>data</highlight> <highlight>to</highlight> <highlight>highlight</highlight> that is different"
          ]
        }
      }
    ]
```

4.  Highlight Failure

```
curl -XPOST 'localhost:9200/test/x/_search?pretty' -d '{
  "from" : 0,
  "size" : 1,
  "query" : {
               "bool" : {
                        "must" : [
                          {
                            "more_like_this" : {
                              "fields" : [
                                "data"
                              ],
                              "like" : [
                                {
                                  "_index" : "test",
                                  "_type" : "x",
                                  "_id" : "1"
                                }
                              ],
                              "max_query_terms" : 50,
                              "min_term_freq" : 1,
                              "min_doc_freq" : 1
                             
                            }
                          }
                        ],
                        "disable_coord" : false,
                        "adjust_pure_negative" : true,
                        "boost" : 1.0
                      }
              },
  "highlight" : {
    "pre_tags" : [
      "<highlight>"
    ],
    "post_tags" : [
      "</highlight>"
    ],
    "number_of_fragments" : 0,
    "type" : "unified",
    "fields" : {
 
      "data" : { }
   
    }
  }

}'
```

Note hit has no highlighting:

```
 "hits" : [
      {
        "_index" : "test",
        "_type" : "x",
        "_id" : "2",
        "_score" : 1.1433042,
        "_source" : {
          "data" : "some data to highlight that is different"
        }
      }
    ]
```

**Comment by javanna** [comment]
@jimczi would you happen to know if more like this query is support by the unified highlighter? Can we do anything to make this work?

**Comment by jimczi** [comment]
`more_like_this` is a special query that relies on index statistics to select the terms that will compose the final query. This operation is done per shard with the main index reader.
Though the highlighters don't use the main index reader, the `fvh` highlighter uses a special reader that is restricted to the term vectors and the `plain` highlighter uses a special index reader that is built from the highlighted documents. For this reason, the highlighting of `more_like_this` query is problematic even with the `plain` highlighter which access index statistics as if the index was composed of a single document. This is why it seems to work with this highlighter but if you set `min_doc_freq` to a value greater than 1 then it returns nothing. 
The solution is not trivial and requires some changes to the extraction of the query terms for highlighters. When `term_vectors` are indexed for the field we should be able to handle `more_like_this` query efficiently so I'll mark this issue as adopt me and I'll try to work on this when I get some time. When the `term_vectors` are not enabled the solution is less trivial and would require some special casing that I am not fund of. I think fixing the `unified` with `term_vectors` should be enough if it's documented correctly.

**Comment by mayya-sharipova** [comment]
Closing, as this is resolved in the latest Elasticsearch version (8.12):

```js
{
  "query": {
    "more_like_this": {
      "fields": [
        "content"
      ],
      "like": [
        {
          "_index": "index1",
          "_id": "doc1"
        }
      ],
      "max_query_terms": 50,
      "min_term_freq": 1,
      "min_doc_freq": 1
    }
  },
  "highlight": {
    "fields": {
      "content": {
        "type": "unified"
      }
    }
  }
}
```

produces the output:

```js
{
  "took": 23,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.2423683,
    "hits": [
      {
        "_index": "index1",
        "_id": "doc2",
        "_score": 1.2423683,
        "_source": {
          "content": "some data to highlight that is very different"
        },
        "highlight": {
          "content": [
            "<em>some</em> <em>data</em> <em>to</em> <em>highlight</em> <em>that</em> <em>is</em> very <em>different</em>"
          ]
        }
      }
    ]
  }
}
```
